### DISCUSSION TITLE
Standardizing `metadata` output

### DISCUSSION THREAD

**Discussion by newsroomdev** [biglocalnews/clean-scraper#58] [Announcements]
**EDITED 08/13**: `case_num` >> `case_id` for consistency

Hello!

As part of our ongoing efforts to maintain consistency and reliability with our data, we need to standardize the `metadata` dictionary that `scrape_meta` generates. Adopting a set specification with tests allows downstream consumers (i.e. scraper orchestration servers and data analysts) to store and retrieve assets from disparate websites.

This is an archival project in many ways. The following proposal aims to codify existing patterns while adapting to various websites contributors encounter.

## Background

**Goal:** Streamline contributing and add more informative pull request checks. By adopting specific `metadata` key-value pairs, separate codebases can read the information and handle scraper orchestration and downloading.

The first step is to streamline contribution by creating a `Site` class with a public `scrape_meta` method. Please see https://github.com/biglocalnews/clean-scraper/blob/dev/docs/decisions/00-deprecate-scrape.md for more information. This method produces a Python list of dictionaries. Each dictionary contains shared, required information and additional distinctive information.

## Proposal

Here is an example of expected JSON output going forward.

**Required**:
- `asset_url` to download the respective assets
- `case_id` to properly store the asset

Additional information, like `parent_page` is saved via a persistent storage server for bookkeeping. Please use `details` for further information to record.

```json
[
    {
        "asset_url": "https://example.com/asset.wav",
        "case_id": "abc123",
        "name": "example_asset.wav",
        "parent_page": "agency/case_id/page.html",
        "title": "Example Title",
        "details": {
            "file_size": 9999,
            "date_modified": "2024-01-01T19:20:00+1:00"
            ...
        }
    },
]
```

### Field Descriptions:

- **required** `asset_url (str)`: The URL to the asset. This is the most important field, ensuring we know what to download.
- **required** `case_id (str)`: A unique identifier for the case.
- `name (str)`: The asset's file name.
- `parent_page (str)`: The asset's parent page file path.
- `title (str)`: A title for the asset.
- `details (dict)`: An object containing additional details. _Examples:_
  - file_size: The size of the file in bytes. (Useful for seeing if the file changed between runs)
  - date_modified: The date and time when the file was last modified.

## Why?

Standardizing helps in several ways:

- Consistency: Ensures that all metadata objects are uniform, making it easier to process them programmatically.
- Reliability: Reduces the chances of errors and inconsistencies in the data. We strive to 
- Efficiency: Simplifies integrating new data sources and maintaining existing ones.

## Validation

If this proposal is adopted, pull request checks can test additional Site modules for the `scrape_meta` method and whether it returns a list of dictionaries that include `asset_url` and `case_id`. Additional options can be passed to these tests to avoid long-running GitHub Actions tasks while testing `scrape_meta`

Currently, contributors can check types before opening pull requests via pre-commits and/or VSCode plugins

### VSCode plugins

Here are a few that I use. If adopted, we can merge in workplace settings with these plugins

```sh
code --install-extension ms-python.python && \
code --install-extension ms-python.black-formatter && \
code --install-extension ms-python.isort && \
code --install-extension ms-python.mypy-type-checker && \
code --install-extension ms-python.vscode-pylance
```

## Feedback/Comments

* Review the proposed schema and provide feedback.
* Update your contributions to adopt this standard.
* Use the provided validation tools to check code.

Thank you for your help in scaling up `clean-scraper`!

Best regards,
Gerald Rich
@biglocalnews

**Comment by newsroomdev** [comment]
cc @zstumgoren @tarakc02

**Comment by stucka** [comment]
I believe @newsroomdev said @pickoffwhite said some assets may belong to multiple cases; in that case, perhaps the case_num should be a list, and the first entry should be used as the definitive location for saving ... ? (If "first" we should explain ... how "first" is determined.)

`filesize` under details seems to miss the underscoring we're using elsewhere. `file_size` might be more appropriate.

We have no metadata schema for case information. At an initial pass, I lumped all that into the metatdata details section, which is rather repetitive but not the end of the world. If we go that way, I would recommend using a "case_" prefix for those variables such that it's clear whether things like `date_modified` refers to the asset or the case.

I don't know if those other "details" sections should be standardized in some way.

We should be clear about whether `date_modified` refers to the original asset's format (which makes comparisons perhaps a bit faster) or whether we're standardizing on what looks like ISO 8601.

**Reply by newsroomdev** [reply]
1. Using the same asset for different case numbers is an edge case that must be evaluated more thoroughly. Is the same asset using different URLs? How many instances has this occurred? Does this affect the 90th percentile of assets? Currently, each metadata dictionary is representative of an anchor link `<a>` on an agency page; multiple anchors to the same link would result in multiple `metadata` dictionaries with different `case_id` values and potentially the same or different `asset_url` strings.
2. Updated `filesize` >> `file_size` ✅ 
3. `details` is optional, allowing for flexibility and variation between sites while capturing information, and would require additional discussion and development after required standardization is checked. **Note:** Contributors are strongly encouraged to include type hinting and tests in their pull requests to assist other contributors in meeting spec 🙏

**Comment by tarakc02** [comment]
- regarding case-numbers, it's great to grab those if they're readily available on the site we're scraping from, but our code for processing and organizing files (that we'll still be running these files through) has logic for case number extraction from file content, so don't know if this should be required? and this goes to mike's question above about multiple cases also -- we want to let the downstream processing code figure these things out.
- what do we do if there is no static `asset_url` (e.g. pages with dynamically loaded content)

**Reply by newsroomdev** [reply]
1. `case_num` is used for grouping downloaded assets. The file contents may have a case number, but `case_num` helps `clean-prefect` download additional video, audio, and photo files to the exact subdirectory folder. I'll follow up on the original thread about multiple cases and tag you.
2. Do you have a specific instance or count of how many times we have been unable to get an `asset_url`? A site scraper should be written to handle each site, so if it's dynamic, it would need to use a more dynamic scraping library.

**Reply by tarakc02** [reply]
moving [this](https://github.com/biglocalnews/clean-scraper/pull/52#issuecomment-2266217631) over to this discussion: "Basically, if the page is dynamic, the scraper would need to handle the site dynamically" -- understood, am wondering since the `asset_url` is a required metadata field, does that mean we should have `"asset_url": null` in the json?

**Reply by tarakc02** [reply]
woops i wrote at the same time as you did. the question about blank asset urls came up because of fullerton pd, which has dynamically loaded data, and the PR #52 contains code to do the dynamic scraping. it does download the data, but there is no usable asset url i don't think...

**Reply by newsroomdev** [reply]
This might be a good question for @stucka who was looking into Next Request sites

**Reply by naumansharifwork** [reply]
@newsroomdev in this #52 it does not have a asset url but it does have a asset id which is used while dynamic scraping so should we keep the asset_id or should we put the id into asset_url?

**Reply by stucka** [reply]
I poked at Fullerton a bit. Take all this with a grain of salt: I'm convinced it's truly static, in that I suspect but have not confirmed the right header stuff can get you in consistently. However, it's also clearly not a "here's a link to this PDF" kind of site -- in fact, they seem to be not actually saving PDFs but are rather dynamically generating them maybe off off images, which is ... a choice. WAVs at least were easy. And dear God the PDFs handling is so not performant.

If (and that's a big if) I'm right this is a gray area -- there may be consistent ways to point to a static point to get the asset in a predictable way but it would require a handler, in much the same way as a YouTube link does.

**Reply by naumansharifwork** [reply]
Hey @stucka i did some investigation on this and you are right, but the pdf in this case is in the form of Images small images like in blocks, so we have to make a logic to combine all and group to make a complete image first and then to convert it into a pdf. 

But the question is we will get the image links, "https://portal.laserfiche.com/Portal/Helper/TileData.aspx?repo=r-3261686e&docID=1538393&x=6&y=0&pageNum=2&scale=6858&ro=0&time=1722713561446&showAnn=1&pageID=3902740&search=" like this but we still wont get the complete asset link to download the pdf in our meta file, either we need to add all the small blocks urls in the meta file and construct the pdf in the scrape phase OR in the case of the pdf file we have to send the document id in our meta and do the next steps (downloading all the relative images and constructing the pdf) in scrape phase.

**Reply by zstumgoren** [reply]
@naumansharifwork Sorry to chime in late, but in order to keep `scrape_meta` as simple as possible, I would vote for adding "all the small block urls in the meta file and construct the pdf in the scrape phase". If you want to build a reference `scrape` method that performs the latter step, we could either use that directly in our downstream orchestration code or mimic its behavior to perform the job at scale. It might also be worth creating a wrapper class that is capable of performing this file stitching from a list of "small block urls" in the metadata JSON -- e.g. a FullteronPDF class in addition to the standard Site class, which the Site class can then make use of. But I'll defer to @newsroomdev on the best place to store such info in the evolving data spec, and to @stucka for thoughts on best generic approach since he's taking a look at re-usability across sites for us

**Reply by zstumgoren** [reply]
> * regarding case-numbers, it's great to grab those if they're readily available on the site we're scraping from, but our code for processing and organizing files (that we'll still be running these files through) has logic for case number extraction from file content, so don't know if this should be required? and this goes to mike's question above about multiple cases also -- we want to let the downstream processing code figure these things out.

@tarakc02 Yep, we're generally favoring organizing files by Case Number -- aka the actual number used by police to track the case -- when it's readily available in the HTML or filename. We figured that approach would be a sensible way to organize downloaded files (and include in the metadata JSON we're producing) when it's easy to get and there's no ambiguity. But it's mainly just a pragmatic approach that we know won't always work -- e.g. San Diego PD uses what they call a Case ID (date + street address) to organize case files, with no obvious references in the source code or file name to the actual Case Number used by police for internal tacking.  In those scenarios, we'll simply use some sensible organizational strategy.

The goal is to hopefully to simplify the downstream work for AutoFolio and human researchers when it's an easy lift.

**Reply by stucka** [reply]
FWIW, I can almost guarantee some case numbers will need to be slugified (e.g., "2018:564") for file system compatibility, and we do not have a slugify routine standardized within `utils` yet. Might not be bad to specify that those case numbers should be slugified when it's possible ... ?

**Reply by stucka** [reply]
To @naumansharifwork 's question about how to handle those multi-image non-PDFs ... my inclination would be that, yeah, we should build a handler in the same way that YouTube videos need a handler. I can work on that if y'all want to pull me off other stuff.

For example, this URL seems to work consistently across browsers without cookies or anything: https://portal.laserfiche.com/Portal/DocView.aspx?id=1502660&repo=r-3261686e , with a simple GET request.

The two variables in that URL get thrown into a payload to /Portal/DocumentService.aspx/GetBasicDocumentInfo ... which returns JSON with the total number of pages, among other useful stuff. So I think if we grabbed that URL as the asset URL, we could absolutely work with it. The page tiling seems, um, annoying but we can probably figure it out and seems to be based around predictable GET requests. The docID is in that GetBasicDocumentInfo, as is the size. The time variable is probably just the epoch time and probably not actually necessary,.

**Reply by naumansharifwork** [reply]
@stucka got it I will change the approach to this, so we will use this link as a asset url and handle the case in the scrape phase to construct the pdf out of it.

**Reply by stucka** [reply]
I spent entirely too much time trying to see if there was some way to access the native image file but I think that can be done only through the API, and API keys only go to clients.

For audio, at least, there's a simple download schema, e.g.: https://portal.laserfiche.com/Portal/ElectronicFile.aspx?docid=1504576&repo=r-3261686e&mode=Export  ...

**Reply by naumansharifwork** [reply]
Hey @stucka so shall we keep the pdf download as it is and change the audio and video to use direct link, or shall we make a function to first get all the image block links in meta phase and construct the pdf from them in scrape phase?

**Reply by stucka** [reply]
@naumansharifwork Above my paygrade and I may be misunderstanding intent someone's intent here -- I've asked @newsroomdev to look.

**Reply by newsroomdev** [reply]
hey @naumansharifwork! i think the URL @stucka [mentioned earlier](https://github.com/biglocalnews/clean-scraper/discussions/58#discussioncomment-10269031), `https://portal.laserfiche.com/Portal/DocView.aspx?id=1502660&repo=r-3261686e` will suffice as the `asset_url` for the `scrape_meta` method and then we can devise a solution for downloading it in the application layer. i'll make a ticket for it, but if there's any additional information that can help us interact with the Laserfiche that's worthwhile to include in `details`

**Comment by tarakc02** [comment]
In some cases, especially with smaller departments, the site will mention that the case is being investigated by a different agency. Nauman highlighted the example of [Fremont PD](https://www.fremontpolice.gov/about-us/transparency-portal/officer-involved-shootings), see the cases for 2018. In these cases, should we include something like:

```
"related_agencies": "XYZ PD"
```

As one of the details? It would be connected to each asset. It's not directly connected to the scraping/archival, but might aid downstream file/case organization work.

**Reply by stucka** [reply]
@tarakc02 If you have that, I think it'd be superb to include it!

I drafted something this morning -- I have even less of an idea if any of this will ever come into use, but, what the heck, can't hurt. I don't even know if these fields are all found in other data sets using the system, or if other datasets will have different variables:
```
    for item in ["created_at", "description", "redacted_at", "doc_date", "highlights"]:
        if item in entry:
            line['details'][item] = entry[item]
```

**Reply by newsroomdev** [reply]
That sounds like a good candidate for the nested `metadata.details` dictionary

**Reply by stucka** [reply]
@newsroomdev If I may suggest something ... any project-specific details can go into details with whatever native names and values are available ... except perhaps for a set of standardized names and values where available, e.g., asset_created and asset_modified in ISO 8601 datetime format, or whatever. So in my example, created_at would remain with the original value, but I'd be encouraged to build out an asset_created field with the converted date.
